### PR TITLE
feat: HTTP payload capturing with Apache HTTP Client library

### PR DESCRIPTION
**Pull request by kenfinnigan**

### CODE REVIEW

File: instrumentation/apache-httpclient/apache-httpclient-5.0/build.gradle [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+dependencies {
+  compileOnly project(":bootstrap")
+  compileOnly("org.apache.httpcomponents.client5:httpclient5:5.0")
+
+  compileOnly("io.opentelemetry:opentelemetry-sdk-extension-autoconfigure")
+  compileOnly("io.opentelemetry.instrumentation:opentelemetry-instrumentation-api-semconv")
+  compileOnly("io.opentelemetry.javaagent:opentelemetry-javaagent-tooling")
+
+  // Needs to be "implementation" for ByteBuddy to find a class we instrument
+  implementation("io.opentelemetry.javaagent.instrumentation:opentelemetry-javaagent-apache-httpclient-5.0:${versions.opentelemetryJavaagentAlpha}")
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/build.gradle:20]: Does this mean we would include in our classloader _another_ copy of `opentelemetry-javaagent-apache-httpclient-5.0`? That'd be probably very bad.

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumenter.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpClientAttributesExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpClientMetrics;
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpSpanNameExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.http.HttpSpanStatusExtractor;
+import io.opentelemetry.instrumentation.api.instrumenter.net.PeerServiceAttributesExtractor;
+import io.opentelemetry.javaagent.bootstrap.internal.CommonConfig;
+import org.apache.hc.core5.http.HttpRequest;
+import org.apache.hc.core5.http.HttpResponse;
+
+public final class ApacheHttpClientInstrumenter {
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumenter.java:32]: I am wondering if we should not instead to like we do in the JS distro and manage the activation of this instrumentation on our own, using the `.addAttributesExtractor` facility in our init code.

**Comment by kenfinnigan** [comment]
This PR now depends on https://github.com/lumigo-io/opentelemetry-java-distro/pull/49 being merged first!

File: utils/src/main/java/io/lumigo/javaagent/utils/Strings.java [hunk 3]
```diff
@@ -25,7 +25,7 @@ public static boolean isBlank(String s) {
       return true;
     }
     for (int i = 0; i < strLen; i++) {
-      if ((Character.isWhitespace(s.charAt(i)) == false)) {
+      if ((!Character.isWhitespace(s.charAt(i)))) {
```
Comment (mmanciop) [inline utils/src/main/java/io/lumigo/javaagent/utils/Strings.java:28]: There's one level of brackets too many

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+                trace
+                    .hasSize(1)
+                    .hasSpansSatisfyingExactly(
+                        span -> span
+                            .hasName("GET")
+                            .hasKind(SpanKind.CLIENT)
+                            .hasAttribute(AttributeKey.stringKey("http.method"), "GET")
+                            .hasAttribute(AttributeKey.stringArrayKey("http.request.header.content_type"),
+                                List.of("application/json"))
+                            .hasAttribute(AttributeKey.stringKey("http.request.body"), "null")
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:197]: To make sure we can distinguish `"null"` and `null`, please add a test showing that a response with the `null` as body (i.e., 4 charactersm content-type text/plain) results in the `"\"null\""` string.

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                            .hasAttribute(AttributeKey.stringKey("http.response.body"), "")
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:198]: I am not sure it makes sense to return the empty string for responses of length zero.

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        .hasSize(1)
+        .hasTracesSatisfyingExactly(
+            trace ->
+                trace
+                    .hasSize(1)
+                    .hasSpansSatisfyingExactly(
+                        span -> span
+                            .hasName("GET")
+                            .hasKind(SpanKind.CLIENT)
+                            .hasAttribute(AttributeKey.stringKey("http.method"), "GET")
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:194]: Is there an OTEL enumeration we can use here to match HTTP methods?

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                            .hasAttribute(AttributeKey.stringArrayKey("http.request.header.content_type"),
+                                List.of("application/json"))
+                            .hasAttribute(AttributeKey.stringKey("http.request.body"), "null")
+                            .hasAttribute(AttributeKey.stringKey("http.response.body"), "")
+                            .hasAttribute(AttributeKey.stringArrayKey("http.response.header.transfer_encoding"),
+                                List.of("chunked"))
+                            .hasAttribute(AttributeKey.longKey("http.status_code"), 200L)
+                            .hasStatus(StatusData.unset())
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:125]: This is not great for the instrumentation to do :-)
Status should be `OK`.

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        .filter(header -> header.getName().equals("Content-Encoding"))
+        .findFirst();
+    if (contentEncodingHeader.isPresent() && contentEncodingHeader.get().getValue().equals("gzip")) {
+      bridge.isGzipped = true;
+    }
+  }
+
+  private static void checkForChunked(ResponsePayloadBridge bridge) {
+    Optional<Header> transferEncodingHeader = Arrays.stream(bridge.headers)
+        .filter(header -> header.getName().equals("Transfer-Encoding"))
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:145]: uh-oh, I don't like that case-sensitive equal! Better to check for equality against a lower-cased header name.

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    if (bridge != null) {
+      return bridge.isGzipped;
+    }
+
+    return false;
+  }
+
+  private static void checkForGzip(ResponsePayloadBridge bridge) {
+    Optional<Header> contentEncodingHeader = Arrays.stream(bridge.headers)
+        .filter(header -> header.getName().equals("Content-Encoding"))
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:136]: uh-oh, I don't like that case-sensitive equal! Better to use equalsIgnoreCase

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        .findFirst();
+    if (contentEncodingHeader.isPresent() && contentEncodingHeader.get().getValue().equals("gzip")) {
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:138]: I think it is better to use `equalsIgnoreCase` using `"gzip"` as first member, i.e.:
```
"gzip".equalsIgnoreCase(contentEncodingHeader.get().getValue())
```

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java (hunk 8 above)
Comment (mmanciop) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:145]: equalsIgnoreCase

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+      bridge.isGzipped = true;
+    }
+  }
+
+  private static void checkForChunked(ResponsePayloadBridge bridge) {
+    Optional<Header> transferEncodingHeader = Arrays.stream(bridge.headers)
+        .filter(header -> header.getName().equals("Transfer-Encoding"))
+        .findFirst();
+    if (transferEncodingHeader.isPresent() && transferEncodingHeader.get().getValue().equals("chunked")) {
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:147]: equalsIgnoreCase

**Review by mmanciop — CHANGES_REQUESTED** [review]

File: utils/src/main/java/io/lumigo/javaagent/utils/Strings.java (hunk 3 above)
Comment (kenfinnigan) [reply utils/src/main/java/io/lumigo/javaagent/utils/Strings.java:28]: Not sure how I managed that, thanks

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java (hunk 6 above)
Comment (kenfinnigan) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:194]: There's one for the `AttributeKey`, which I've switched to, but not for the values themselves

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java (hunk 7 above)
Comment (kenfinnigan) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:125]: This is not us, this is what OTeL sets it to in this situation

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java (hunk 8 above)
Comment (kenfinnigan) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:145]: Good call out. Changed

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java (hunk 5 above)
Comment (kenfinnigan) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:198]: I've modified the setting of a response body to only occur if the response is a 200. It seems reasonable, but open to other suggestions

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentation.java [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+        @Advice.Argument(0) ClassicHttpRequest request,
+        @Advice.Return Object result,
+        @Advice.Thrown Throwable throwable,
+        @Advice.Local("otelContext") Context context,
+        @Advice.Local("otelScope") Scope scope) {
+      if (scope == null) {
+        return;
+      }
+
+      scope.close();
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentation.java:192]: Can closing the scope fail? If so, should put `ApacheHttpClientHelper.doMethodExit` in a `finally`?

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java (hunk 7 above)
Comment (mmanciop) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:130]: Yes and I mildly disapprove of it

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java [hunk 13]
```diff
@@ -133,18 +133,18 @@ public static boolean isGzipped(Context context) {
...
-    if (contentEncodingHeader.isPresent() && contentEncodingHeader.get().getValue().equals("gzip")) {
+    if (contentEncodingHeader.isPresent() && "gzip".equalsIgnoreCase(contentEncodingHeader.get().getValue())) {
       bridge.isGzipped = true;
     }
   }
 
   private static void checkForChunked(ResponsePayloadBridge bridge) {
     Optional<Header> transferEncodingHeader = Arrays.stream(bridge.headers)
-        .filter(header -> header.getName().equals("Transfer-Encoding"))
+        .filter(header -> header.getName().equalsIgnoreCase("transfer-encoding"))
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:145]: The other direction `<constant>.equals[IgnoreCase](<nullable thing>)` is 100% better :-)

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java [hunk 14]
```diff
@@ -133,18 +133,18 @@ public static boolean isGzipped(Context context) {
 
   private static void checkForGzip(ResponsePayloadBridge bridge) {
     Optional<Header> contentEncodingHeader = Arrays.stream(bridge.headers)
-        .filter(header -> header.getName().equals("Content-Encoding"))
+        .filter(header -> header.getName().equalsIgnoreCase("content-encoding"))
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:136]: The other direction <constant>.equals[IgnoreCase](<nullable thing>) is 100% better :-)

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java [hunk 15]
```diff
@@ -75,7 +80,7 @@ void testConnectionReset() {
                       span -> span
                           .hasName("GET")
                           .hasKind(SpanKind.CLIENT)
-                          .hasAttribute(AttributeKey.stringKey("http.method"), "GET")
+                          .hasAttribute(SemanticAttributes.HTTP_METHOD, "GET")
                           .hasAttribute(AttributeKey.stringArrayKey("http.request.header.content_type"),
```
Comment (mmanciop) [inline instrumentation/apache-httpclient/apache-httpclient-5.0/src/test/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ConnectionTest.java:84]: This also has an OTEL semantic convention, no?

**Review by mmanciop — COMMENTED** [review]
LGTM, but before merging, we need regression testing on the various versions of Apache HTTP client 5.x :-)

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentation.java (hunk 12 above)
Comment (kenfinnigan) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ApacheHttpClientInstrumentation.java:192]: No. If it finds the scope is not part of the current context, it logs a "fine" message saying close is ignored

File: instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java (hunk 13 above)
Comment (kenfinnigan) [reply instrumentation/apache-httpclient/apache-httpclient-5.0/src/main/java/io/opentelemetry/javaagent/instrumentation/apachehttpclient/v5_0/ResponsePayloadBridge.java:145]: Apologies, not sure why I didn't just do that yesterday

**Review by GuyMoses — DISMISSED** [review]
